server/src/instant/db/model/attr.clj

```diff
@@ -681,59 +656,82 @@
 ;; Creates a wrapper over attrs. Makes them act like a regular list, but
 ;; we can also index them on demand so that our access patterns will be
 ;; efficient.
-(deftype Attrs [elements cache]
+(declare wrap-attrs)
+
+(deftype Attrs [elements
+                by-id-cache
+                by-fwd-ident-cache
+                by-rev-ident-cache
+                ids-by-etype-cache
+                ea-ids-by-etype-cache]
   clojure.lang.ISeq
   (count [_this]
     (count elements))
   (first [_this]
     (first elements))
   (next [_this]
-    (let [nxt (next elements)]
-      (if nxt
-        (Attrs. nxt (delay (index-attrs nxt)))
-        nil)))
+    (some-> elements next wrap-attrs))
   (more [_this]
-    (if-let [nxt (next elements)]
-      (Attrs. nxt (delay (index-attrs nxt)))
+    (if-some [nxt (next elements)]
+      (wrap-attrs nxt)
       clojure.lang.PersistentList/EMPTY))
   (empty [_this]
-    (Attrs. () (delay {})))
+    (wrap-attrs ()))
   (equiv [_this other]
     (= elements other))
   (cons [_this o]
-    (let [new-elements (cons o elements)]
-      (Attrs. new-elements (delay (index-attrs new-elements)))))
+    (wrap-attrs (cons o elements)))
   (seq [this]
     (if (empty? elements)
       nil
       this))
 
+  clojure.lang.IHashEq
+  (hasheq [_this]
+    (clojure.lang.Murmur3/hashUnordered elements))
+
+  java.lang.Object
+  (equals [_this o]
+    (and (instance? Attrs o)
+         (= (set elements)
+            (set (.-elements ^Attrs o)))))
+
   AttrsExtension
   (seekById [_this id]
-    (-> @cache
-        :by-id
-        (get id)))
+    (get @by-id-cache id))
   (seekByFwdIdentName [_this fwdIdent]
-    (-> @cache
-        :by-fwd-ident
-        (get fwdIdent)))
+    (get @by-fwd-ident-cache fwdIdent))
   (seekByRevIdentName [_this revIdent]
-    (-> @cache
-        :by-rev-ident
-        (get revIdent)))
+    (get @by-rev-ident-cache revIdent))
   (attrIdsForEtype [_this etype]
-    (-> @cache
-        :ids-by-etype
-        (get etype #{})))
-  (unwrap [_this]
-    elements)
+    (get @ids-by-etype-cache etype #{}))
   (eaIdsForEtype [_this etype]
-    (-> @cache
-        :ea-ids-by-etype
-        (get etype #{}))))
+    (get @ea-ids-by-etype-cache etype #{}))
+  (unwrap [_this]
+    elements))
 
 (defn wrap-attrs [attrs]
-  (Attrs. attrs (delay (index-attrs attrs))))
+  (Attrs.
+   attrs
+   ;; by-id-cache
+   (delay
```

Comment:
My thinking was that we'd save cpu cycles in the long run by building all of the caches we'd use in one scan of the attrs, even if we populated caches that we didn't use. Are you sure multiple traversals over the attrs isn't going to perform worse?

Reply:
Yes. The time is dominated by assoc anyways. We are actually faster after my changes, 35 µs → 30 µs to deref everything